cmd/apply-crds/main.go

```diff
@@ -0,0 +1,163 @@
+/*
+2024 NVIDIA CORPORATION & AFFILIATES
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/retry"
+)
+
+var (
+	crdsDir string
+)
+
+func initFlags() {
+	flag.StringVar(&crdsDir, "crds-dir", "", "Path to the directory containing the CRD manifests")
+	flag.Parse()
+
+	if crdsDir == "" {
+		log.Fatalf("CRDs directory is required")
+	}
+
+	if _, err := os.Stat(crdsDir); os.IsNotExist(err) {
+		log.Fatalf("CRDs directory %s does not exist", crdsDir)
+	}
+}
+
+func main() {
+	ctx := context.Background()
+
+	initFlags()
+
+	config, err := getKubeConfig()
+	if err != nil {
+		log.Fatalf("Failed to get Kubernetes config: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Failed to create API extensions client: %v", err)
+	}
+
+	// Walk the directory recursively and apply each YAML file.
+	err = filepath.Walk(crdsDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() || filepath.Ext(path) != ".yaml" {
+			return nil
+		}
+
+		log.Printf("Apply CRDs from file: %s", path)
+		if err := applyCRDsFromFile(ctx, client, path); err != nil {
+			log.Printf("Failed to apply CRD %s: %v", path, err)
+		}
+		return nil
+	})
+	if err != nil {
+		log.Fatalf("Error walking the path %s: %v", crdsDir, err)
+	}
+}
+
+// applyCRDsFromFile reads a YAML file, splits it into documents, and applies each CRD to the cluster.
+func applyCRDsFromFile(ctx context.Context, client *clientset.Clientset, filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("open file %q: %v", filePath, err)
+	}
+	defer file.Close()
+
+	// Create a decoder that reads multiple YAML documents.
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 4096)
```

Comment:
I believe using a `reader` is more efficient.

```go
	reader := yaml.NewYAMLReader(bufio.NewReader(file))

	var crdsToApply []*apiextensionsv1.CustomResourceDefinition
	for {
		doc, err := reader.Read()
		if err != nil {
			if err == io.EOF {
				break
			}
		}
		crd := &apiextensionsv1.CustomResourceDefinition{}
		decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(doc), len(doc))
		if err := decoder.Decode(crd); err != nil {
			return fmt.Errorf("decode YAML: %v", err)
		}
		crdsToApply = append(crdsToApply, crd)
	}
```

`doc, err := reader.Read()`  this will give you a full Yaml document instead of limiting it to the buffer size.

Reply:
Good point, will implement it 👍🏻 thanks

Reply:
Tested it locally with 20.000 CRD updates. Both took about 44s, so no performance optimization unfortunately.
I will keep it as is